### PR TITLE
Add arg_choice arg to DeclareLaunchArguments

### PR DESCRIPTION
**Pull request by v-lopez**
Fixes https://github.com/ros2/launch/issues/481

### CODE REVIEW

File: launch/launch/actions/declare_launch_argument.py [hunk 1]
```diff
@@ -80,6 +82,7 @@ def __init__(
         *,
         default_value: Optional[SomeSubstitutionsType] = None,
         description: Text = 'no description given',
+        arg_choices: List[Text] = None,
```
Comment (hidmic) [inline launch/launch/actions/declare_launch_argument.py:85]: @v-lopez nit: why not simply `choices`?

File: launch/launch/actions/declare_launch_argument.py [hunk 2]
```diff
@@ -90,7 +93,10 @@ def __init__(
         else:
             self.__default_value = normalize_to_list_of_substitutions(default_value)
         self.__description = description
+        if arg_choices is not None:
+            self.__description += " Valid choices are: " + str(arg_choices)
```
Comment (hidmic) [inline launch/launch/actions/declare_launch_argument.py:97]: @v-lopez +1 to list valid choices in the description, but `self.__description` may actually be `'no description given'`. Perhaps `description` should be an `Optional` and be better handled here?

File: launch/launch/actions/declare_launch_argument.py [hunk 3]
```diff
@@ -90,7 +93,10 @@ def __init__(
         else:
             self.__default_value = normalize_to_list_of_substitutions(default_value)
```
Comment (hidmic) [inline launch/launch/actions/declare_launch_argument.py:94]: @v-lopez it'd be good to flag inconsistencies between `choices` and the `default_value` as early as possible.

File: launch/launch/actions/declare_launch_argument.py [hunk 4]
```diff
@@ -47,11 +47,13 @@ class DeclareLaunchArgument(Action):
     command-line arguments when launched with ``ros2 launch ...``.
 
     In addition to the name, which is also where the argument result is stored,
-    launch arguments may have a default value and a description.
+    launch arguments may have a default value, a choice list and a description.
```
Comment (hidmic) [inline launch/launch/actions/declare_launch_argument.py:50]: @v-lopez nit: 
```suggestion
    launch arguments may have a default value, a list of valid value choices, and a description.
```

File: launch/launch/actions/declare_launch_argument.py [hunk 5]
```diff
@@ -131,6 +140,11 @@ def description(self) -> Text:
         """Getter for self.__description."""
         return self.__description
 
+    @property
+    def arg_choices(self) -> Text:
```
Comment (hidmic) [inline launch/launch/actions/declare_launch_argument.py:144]: @v-lopez 
```suggestion
    def arg_choices(self) -> List[Text]:
```
?

File: launch/launch/actions/declare_launch_argument.py (hunk 1 above)
Comment (v-lopez) [reply launch/launch/actions/declare_launch_argument.py:85]: Sure, fixing it.

File: launch/launch/actions/declare_launch_argument.py (hunk 3 above)
Comment (v-lopez) [reply launch/launch/actions/declare_launch_argument.py:94]: Agreed, but I'm a little bit lost regarding the substitution handling.

Would a valid approach be:
1. Check if default_value is not an instance of `Substitution`
2. If it is not, raise an error if the `default_value` is not in `choices`

File: launch/launch/actions/declare_launch_argument.py (hunk 2 above)
Comment (v-lopez) [reply launch/launch/actions/declare_launch_argument.py:97]: Ok, which message would you expect if no `description` is provided, but `choices` is?

File: launch/launch/actions/declare_launch_argument.py (hunk 3 above)
Comment (hidmic) [reply launch/launch/actions/declare_launch_argument.py:95]: Correct.

File: launch/launch/actions/declare_launch_argument.py [hunk 6]
```diff
@@ -79,19 +81,41 @@ def __init__(
...
+                self.__description = 'no description given'
+            else:
+                self.__description = 'One of: ' + str(choices)
+        else:
+            self.__description = description
+            if choices is not None:
+                self.__description += " Valid choices are: " + str(choices)
 
-        self.__logger = launch.logging.get_logger(__name__)
+        self.__choices = choices
```
Comment (hidmic) [inline launch/launch/actions/declare_launch_argument.py:118]: @v-lopez meta: shall we check that `choices` is not an empty list as well?

**Review by hidmic — APPROVED** [review]
Overall LGTM

File: launch/launch/actions/declare_launch_argument.py (hunk 2 above)
Comment (hidmic) [reply launch/launch/actions/declare_launch_argument.py]: The one you proposed looks fine.

**Comment by hidmic** [comment]
CI up to `launch` (and `test_launch_ros`, JIC):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13491/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8398/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11213/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13556/)

File: launch/launch/actions/declare_launch_argument.py (hunk 6 above)
Comment (v-lopez) [reply launch/launch/actions/declare_launch_argument.py:124]: Sure, please check https://github.com/ros2/launch/pull/483/commits/20797ba791c818e6ae86d2f30f6f642e9933d90c.

File: launch/launch/actions/declare_launch_argument.py [hunk 7]
```diff
@@ -79,19 +81,47 @@ def __init__(
...
         super().__init__(**kwargs)
         self.__name = name
+        self.__logger = launch.logging.get_logger(__name__)
         if default_value is None:
             self.__default_value = default_value
         else:
             self.__default_value = normalize_to_list_of_substitutions(default_value)
-        self.__description = description
+        if choices is not None:
+            if choices == []:
```
Comment (hidmic) [inline launch/launch/actions/declare_launch_argument.py:97]: @v-lopez 
```suggestion
            if not choices:
```
? Or 
```suggestion
            if len(choices) == 0:
```
 would do as well.

File: launch/test/launch/actions/test_declare_launch_argument.py [hunk 8]
```diff
@@ -57,3 +71,15 @@ def test_declare_launch_argument_execute():
     lc2 = LaunchContext()
     assert action2.visit(lc2) is None
     assert lc1.launch_configurations['name'] == 'value'
+
+    action3 = DeclareLaunchArgument('name', default_value="var1", choices=["var1", "var2"])
```
Comment (hidmic) [inline launch/test/launch/actions/test_declare_launch_argument.py:75]: @v-lopez nit: use single quotes, or flake8 will complain.

File: launch/launch/actions/declare_launch_argument.py (hunk 7 above)
Comment (v-lopez) [reply launch/launch/actions/declare_launch_argument.py:97]: Going with `len`, I want it to throw if for any reason we change the code and choices is `None` and reaches this statement.

File: launch/test/launch/actions/test_declare_launch_argument.py (hunk 8 above)
Comment (v-lopez) [reply launch/test/launch/actions/test_declare_launch_argument.py]: Done

**Review by hidmic — APPROVED** [review]
CI up to `launch` (and `test_launch_ros`):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13515/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8418/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11241/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13582/)

**Comment by hidmic** [comment]
@v-lopez please, check tests are passing locally before submission.

**Comment by v-lopez** [comment]
Sorry about that, I pushed more fixes, how do you test this?

I was using `colcon test --packages-select launch` and running also ament_flake8 and ament_pep257 on the package but did not detect these issues.

**Comment by hidmic** [comment]
An old `flake8` version, perhaps? There's one more to go!

**Comment by hidmic** [comment]
CI up to `launch` (and `test_launch_ros`):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13549/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8443/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11272/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13614/)

**Comment by v-lopez** [comment]
I'm sorry there seems to be yet another problem with flake.. 
It reports: 
```
  ./launch/actions/declare_launch_argument.py:178:21: Q003 Change outer quotes to avoid escaping inner quotes
```
But I believe I cannot change outer quotes or I'll break Q000. I'll try using double quotes inside.

**Comment by hidmic** [comment]
CI up to `launch` (and `test_launch_ros`):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13555/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8449/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11278/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13620/)

**Comment by v-lopez** [comment]
Could this be merged and released on foxy? Got some work pending on this.

**Comment by hidmic** [comment]
It could, yes. Thanks for your contribution @v-lopez. Feel free to open a backport PR if you want this on Foxy.

**Pull request merged**